GH-41558: [C++] Improve fixed_width_test_util.h

llama90 opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

Improve the `fixed_width_test_util.h`.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

- Move the `fixed_width_test_util.h` to `arrow/testing`
- Divide the `fixed_width_test_util` to `.cc` and `.h`
- Remove unused headers

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

Yes

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

No

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* GitHub Issue: #41558

llama90 commented:
@felipecrv Hello. 

Is this what you were thinking? 

I would appreciate it if you could review it when you have time. Thank you.

felipecrv commented on cpp/src/arrow/testing/fixed_width_test_util.h:
```diff
@@ -17,16 +17,6 @@
 
 #pragma once
 
-#include <cstddef>
-#include <cstdint>
-#include <memory>
-#include <vector>
-
-#include "arrow/array/builder_primitive.h"
-#include "arrow/builder.h"
-#include "arrow/type.h"
-#include "arrow/util/checked_cast.h"
-
```
I meant reducing the code in the header. We should keep the necessary includes.

The class should only declare the public function members here and all the private members and template code can be defined in the `.cc` file (to be created) outside of the class.

See how `chunk_resolver.h/cc` does it reducing the bloat in the header: https://github.com/apache/arrow/blob/51689a040cbe3dee8702cd899a33fa62e0616bf1/cpp/src/arrow/chunk_resolver.cc#L44-L45

llama90 replied:
Ah, so I need to organize the code according to C++ conventions. I should include only the declarations in the .h file and put the implementations in the .cc file. 

I will refer to the code you shared, organize it accordingly, and then upload it.

Thank you for your guidance.

felipecrv commented on cpp/src/arrow/testing/fixed_width_test_util.h:
```diff
@@ -35,24 +29,80 @@ class NestedListGenerator {
-  }
+      const std::shared_ptr<DataType>& inner_type, const std::vector<int>& sizes);
 
   /// \brief Create a nested FixedListType.
```
You should move these template functions to an anonymous namespace in the `.cc` file.

felipecrv commented on cpp/src/arrow/testing/fixed_width_test_util.h:
```diff
@@ -120,84 +170,6 @@ class NestedListGenerator {
     }
     return nested_builder->Finish();
   }
-
```
You should have the definition of `NestedListArray` in the `.cc` as well.

llama90 commented:
@felipecrv Hello, and thank you for your review.

I have applied the changes as advised, but I am a bit unsure about one aspect, particularly because I am not very familiar with template functions.

Should the `VisitAllNestedListConfigurations` template function also be moved to an anonymous namespace? It is directly called from the test function (`vector_selection_test.cc`), so I have kept it as it currently is.

Additionally, I made modifications to the `Makefile` to accommodate the Windows CI process. Could you please confirm if this approach is correct?

I would greatly appreciate your advice on these matters.

Thank you!

llama90 commented:
I think I applied everything you suggested:

- Converted the template function to a regular function.
- Modified to not include relative paths in the Makefile (this was a problem due to a missing macro as `ARROW_TESTING_EXPORT`).

Additionally, I want to express my gratitude for your patience and thorough review. It has been instrumental in completing this PR. 

Thank you! cc @felipecrv

felipecrv commented on cpp/src/arrow/testing/fixed_width_test_util.h:
```diff
@@ -0,0 +1,72 @@
+
+#pragma once
+
+#include <vector>
+
+#include "arrow/builder.h"
+#include "arrow/testing/visibility.h"
```
We should try to "include what you use" [1] as much as possible. Even when headers that we need are already included by other headers.

- `<memory>`
- `<functional>`
- `<cstdint>`
- `"arrow/type.h"` for `DataType` (complete type is needed for the pointers and vector declarations)
- `"arrow/type_fwd.h"` forward-declares `ArrayBuilder` (which we don't need to be a complete type in the header because we only pointers/references to it in this header)

With these included, you should be able to remove `"arrow/builder.h"` which is a very expensive header.

[1] https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md

felipecrv commented on cpp/src/arrow/testing/fixed_width_test_util.cc:
```diff
@@ -0,0 +1,176 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include "fixed_width_test_util.h"
```
With the removal of include bloat in the header, you should also add all the includes for things referred from this file:

 - `"arrow/array/builder_primitive.h"` for `NumericBuilder` and `ArrowBuilder`
 - `"arrow/array/builder_base.h"` for `MakeBuilder` [1]
 - `<cstdint>`
 - `<limits>`
 - `<vector>`
 - `<functional>`
 - `"arrow/type.h"`
 
[1] my header had an include to `arrow/builder.h`, but that was unnecessary now that I thought more about it

felipecrv commented on cpp/src/arrow/testing/fixed_width_test_util.cc:
```diff
@@ -0,0 +1,176 @@
+  return Status::OK();
+}
+
+template Status AppendNumeric<Int8Type>(ArrayBuilder* builder, int64_t* next_value);
+template Status AppendNumeric<Int16Type>(ArrayBuilder* builder, int64_t* next_value);
+template Status AppendNumeric<Int32Type>(ArrayBuilder* builder, int64_t* next_value);
+template Status AppendNumeric<Int64Type>(ArrayBuilder* builder, int64_t* next_value);
```
You don't need these (in this context) because since this template is in an anon namespace the compiler will only instantiate the 4 instances that are referred to in this compilation unit.

felipecrv reviewed (changes requested):
> Additionally, I want to express my gratitude for your patience and thorough review. It has been instrumental in completing this PR.

Thank you for not giving up ;) C++ is weird.

I added comments about includes, but other than that everything is looking very good.

llama90 commented:
@felipecrv I applied your suggestions

Thank you for review :)
